kernel/src/engine/arrow_expression/evaluate_expression.rs

```diff
@@ -115,29 +111,63 @@ pub(crate) fn evaluate_expression(
         (Struct(_), _) => Err(Error::generic(
             "Data type is required to evaluate struct expressions",
         )),
-        (Unary(UnaryExpression { op, expr }), _) => {
+        (Predicate(pred), None | Some(&DataType::BOOLEAN)) => {
+            let result = evaluate_predicate(pred, batch)?;
+            Ok(Arc::new(result))
+        }
+        (Predicate(_), Some(data_type)) => Err(Error::generic(format!(
+            "Predicate evaluation produces boolean output, but caller expects {data_type:?}"
+        ))),
+        (Binary(BinaryExpression { op, left, right }), _) => {
+            let left_arr = evaluate_expression(left.as_ref(), batch, None)?;
+            let right_arr = evaluate_expression(right.as_ref(), batch, None)?;
+
+            type Operation = fn(&dyn Datum, &dyn Datum) -> Result<ArrayRef, ArrowError>;
+            let eval: Operation = match op {
+                Plus => add,
+                Minus => sub,
+                Multiply => mul,
+                Divide => div,
+            };
+
+            Ok(eval(&left_arr, &right_arr)?)
+        }
+    }
+}
+
+pub(crate) fn evaluate_predicate(
+    predicate: &Predicate,
+    batch: &RecordBatch,
+) -> DeltaResult<BooleanArray> {
+    use BinaryPredicateOp::*;
+    use Predicate::*;
+    match predicate {
+        BooleanExpression(expr) => {
+            // Grr -- there's no way to cast an `Arc<dyn Array>` back to its native type, so we
+            // can't use `Arc::into_inner` here and must unconditionally clone instead.
+            let arr = evaluate_expression(expr, batch, Some(&DataType::BOOLEAN))?;
+            Ok(downcast_to_bool(&arr)?.clone())
```

Comment:
Gah! I thought I had it with:
```suggestion
            arr.into_data().try_into().map_err(|_| Error::generic("expected boolean array"))
```

We can do `arr.into_data()` because have a blanket `impl Array for Arc<dyn Array>` [here](https://docs.rs/arrow/latest/arrow/array/array/type.ArrayRef.html#impl-Array-for-Arc%3Cdyn+Array%3E)

_BUT_, looking at the [code](https://docs.rs/arrow-array/55.0.0/src/arrow_array/array/mod.rs.html#354) for an `Arc` this just calls `to_data`, so it'll clone :(

I do think this is a little more clean though so maybe we want it anyway? And then we can also remove `downcast_to_bool`.

Reply:
> Problem is, TryFrom<ArrayData> for BooleanArray is just the (infallible) blanket impl based on [From<ArrayData> for BooleanArray](https://docs.rs/arrow/latest/arrow/array/struct.BooleanArray.html#impl-From%3CArrayData%3E-for-BooleanArray), which [panics](https://docs.rs/arrow-array/55.0.0/src/arrow_array/array/boolean_array.rs.html#382) on type mismatch. I couldn't find any fallible version of that code?

Ahh yeah. We could check the conditions of the asserts though, i.e. that the `DataType` of the `ArrayData` is `Boolean` and that there's only one buffer. That would give us confidence that it won't panic (although arrow updates could of course change the conditions).

But anyway it's mostly moot because We end up cloning anyway, and we can't get the inner out of the Arc since it's a trait.